streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingWindowStore.java

```diff
@@ -0,0 +1,694 @@
+        private LRUCacheEntry cachedBaseValue;
+        private final SegmentedCacheFunction cacheFunction;
+
+        private ThreadCache.MemoryLRUCacheBytesIterator current;
+
+        private CacheIteratorWrapper(final Bytes key,
+                                     final long timeFrom,
+                                     final long timeTo,
+                                     final boolean forward,
+                                     final boolean index) {
+            this(key, key, timeFrom, timeTo, forward, index);
+        }
+
+        private CacheIteratorWrapper(final Bytes keyFrom,
+                                     final Bytes keyTo,
+                                     final long timeFrom,
+                                     final long timeTo,
+                                     final boolean forward,
+                                     final boolean index) {
+            this.keyFrom = keyFrom;
+            this.keyTo = keyTo;
+            this.timeTo = timeTo;
+            this.forward = forward;
+            this.useIndex = index;
+
+            cacheFunction = index ? indexKeyCacheFunction : baseKeyCacheFunction;
+
+            this.segmentInterval = cacheFunction.getSegmentInterval();
+
+            if (forward) {
+                this.lastSegmentId = cacheFunction.segmentId(Math.min(timeTo, maxObservedTimestamp.get()));
+                this.currentSegmentId = cacheFunction.segmentId(timeFrom);
+
+                setCacheKeyRange(timeFrom, currentSegmentLastTime());
+                this.current = context.cache().range(cacheName, cacheKeyFrom, cacheKeyTo);
+            } else {
+                this.currentSegmentId = cacheFunction.segmentId(Math.min(timeTo, maxObservedTimestamp.get()));
+                this.lastSegmentId = cacheFunction.segmentId(timeFrom);
+
+                setCacheKeyRange(currentSegmentBeginTime(), Math.min(timeTo, maxObservedTimestamp.get()));
+                this.current = context.cache().reverseRange(cacheName, cacheKeyFrom, cacheKeyTo);
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (current == null) {
+                return false;
+            }
+
+            if (useIndex) {
+                do {
+                    // If iterating from index, need to make sure base key/value exist in cache
+                    while (current.hasNext()) {
+                        final Bytes cacheIndexKey = current.peekNextKey();
+                        final Bytes indexKey = indexKeyCacheFunction.key(cacheIndexKey);
+                        final Bytes baseKey = indexKeyToBaseKey(indexKey);
+                        final Bytes cachedBaseKey = baseKeyCacheFunction.cacheKey(baseKey);
+                        cachedBaseValue = context.cache().get(cacheName, cachedBaseKey);
+                        if (cachedBaseValue != null) {
+                            return true;
+                        }
+                        current.next();
```

Comment:
Ideally, we should delete the index if base is not in cache since this dangling index is useless. But the iterator doesn't have remove api and deleting directly from map can cause concurrent modification error.

Reply:
My understanding is that base value would be not available only if the previous put failed on base but succeeded in index, right? In that case would the index store's size keep increasing over time?

Reply:
base value could be evicted from LRU.

Reply:
NVM I realized that we rely on segment-based cleaning anyways.